Add preference saving from dialog for folders with extensions (#4535)
Napari/napari: docs/release/release_0_4_16.md

- Calling `viewer.open` *without* passing a plugin will result in an error if you have not saved a reader preference for that file pattern *and* multiple plugins can claim the file

>     - To save a preference for a file pattern at the command line:

Reviewer: ```suggestion
    - To save a preference for a file pattern in Python, use:
```

    ```python
    from napari.settings import get_settings
    get_settings().plugins.extension2reader['*.tif'] = napari_tifffile
    get_settings().plugins.extension2reader['*.zarr'] = napari_ome_zarr

>     ```
>     - To pass a plugin in a Python script:

Reviewer: ```suggestion
    ```

    - To specify a plugin in a Python script:
```

    ```python
    import napari

    viewer = napari.Viewer()
    viewer.open('my-path.tif') # this will throw MultipleReaderError if napari_tifffile is installed as both it and builtins could open the file
    viewer.open('my-path.tif', plugin=napari_tifffile) # this won't
    ```

>     - To pass a plugin at the command line

Reviewer: ```suggestion
    - To specify a plugin at the command line, use:
```

    
    ```sh
    napari my-path.tif --plugin napari_tifffile

>     ```
> - A preferred reader failing to read your file will result in an error e.g. if you saved `napari_tifffile` as a preference for TIFFs but then uninstalled it

Reviewer: ```suggestion
    ```

- A preferred reader failing to read your file will result in an error e.g. if you saved `napari_tifffile` as a preference for TIFFs but then uninstalled it
```

I didn't realise this was the case @DragaDoncila and I think it's a big problem, because napari doesn't have per-environment preferences. imho plugin preferences that don't match an installed plugin should be ignored...

Author: Two things I guess, @jni.

- If a preferred plugin is available but just fails to read the file for some reason we raise `ReaderPluginError`, I think you can agree that's not controversial
- If a preferred plugin is missing, we raise `MissingAssociatedReaderError` which is the bit I think you'd like to change. Idk how I feel about that. I actually didn't realize settings weren't per environment (though I should've), but I don't know if ignoring preferences is the right way? I feel like it could be confusing to associate a reader, it's not found, and we just ignore that entirely. Especially since you can't add two preferences for the same file type. Maybe we should warn? Idk I'd love more eyes on this @tlambert03 @nclack? I think if we want to change that pre-release it should be a separate PR anyway.

- Calling `viewer.open` *without* passing a plugin will result in an error if you have not saved a reader preference for that file pattern *and* multiple plugins can claim the file
    - To save a preference for a file pattern in Python, use:

    ```python
    from napari.settings import get_settings

>     get_settings().plugins.extension2reader['*.tif'] = napari_tifffile

Reviewer 2: from the example, it's unclear what the values are here.   is `napari_tifffile` a module?  a string?

Author: thanks for catching that, they're strings. Have updated now.

    get_settings().plugins.extension2reader['*.zarr'] = napari_ome_zarr
    ```

    - To specify a plugin in a Python script:

    ```python
    import napari

    viewer = napari.Viewer()
    viewer.open('my-path.tif') # this will throw MultipleReaderError if napari_tifffile is installed as both it and builtins could open the file

>     viewer.open('my-path.tif', plugin=napari_tifffile) # this won't

Reviewer 2: same here... shouldn't the plugin be a string?

Reviewer 2: > this will throw MultipleReaderError if napari_tifffile is installed as both it and builtins could open the file

apologies if I'm forgetting the conversation we had about this, and I know this is from a previous PR... but we decided against letting builtins take it programatically?  Seems a bummer that installing a plugin that recognizes tiff or png or something now means that a simple programatic `viewer.open('something.tif')` is now a full on error

Author: Well, we didn't decide anything yet @tlambert03. I remember at the hackathon we said maybe `builtins` should take it, but @jni's opinion in #4389 (and I tend to agree), is that if we do this at all, it should be the *other* plugin that takes the file.

Do you think deciding on this is critical before release? Fwiw the change itself would be trivial, it's just a matter of whether we want to make it. The fact that your suggestion is exactly opposite to Juan's makes me think that maybe this is a bit of magic we should just steer away from? I totally get the reticence and the feeling of "I just installed the plugin obviously I want to use it", but I just don't know if that's enough reason to sprinkle a bit of the ol' magic back in.
